File: src/services/message-broker/message-broker-factory.ts
/**
 * Message Broker 팩토리.
 *
 * 환경 변수 QUEUE_SERVICE_TYPE에 따라 적절한 MessageBroker 어댑터를 생성한다.
 * SPEC.md §4.1 참조.
 */

import { AzureKeyCredential } from '@azure/eventgrid';
import type { Settings } from '../../shared/config';
import { getLogger } from '../../shared/logger';
import { EventGridAdapter } from './event-grid-adapter';
import type { MessageBroker } from './message-broker';

const logger = getLogger('message-broker-factory');

const SUPPORTED_BROKER_TYPES = new Set(['EVENT_GRID']);

export class MessageBrokerFactory {
  static create(settings: Settings): MessageBroker {
    const brokerType = settings.QUEUE_SERVICE_TYPE.toUpperCase();

    if (brokerType === 'EVENT_GRID') {
      const endpoint = settings.EVENT_GRID_TOPIC_ENDPOINT;
      const key = settings.EVENT_GRID_TOPIC_KEY;

      if (!endpoint || !key) {
        logger.error('EVENT_GRID_TOPIC_ENDPOINT 또는 EVENT_GRID_TOPIC_KEY 미설정', {
          has_endpoint: !!endpoint,
          has_key: !!key,
        });
        throw new Error(
          'QUEUE_SERVICE_TYPE=EVENT_GRID이지만 EVENT_GRID_TOPIC_ENDPOINT/EVENT_GRID_TOPIC_KEY가 설정되지 않았습니다.',
        );
      }

      const credential = new AzureKeyCredential(key);
      return new EventGridAdapter(endpoint, credential);
    }

    const supported = [...SUPPORTED_BROKER_TYPES].sort().join(', ');
    throw new Error(
      `지원하지 않는 QUEUE_SERVICE_TYPE: '${settings.QUEUE_SERVICE_TYPE}'. 지원: ${supported}`,
    );
  }
}

File: src/services/message_broker/message_broker_factory.py
"""Message Broker 팩토리.

환경 변수 QUEUE_SERVICE_TYPE에 따라 적절한 MessageBroker 어댑터를 생성한다.

SPEC.md §4.1 참조.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from azure.core.credentials import AzureKeyCredential

from src.services.message_broker.event_grid_adapter import EventGridAdapter

if TYPE_CHECKING:
    from src.services.message_broker.message_broker import MessageBroker
    from src.shared.config import Settings

# 지원하는 브로커 타입
SUPPORTED_BROKER_TYPES = {"EVENT_GRID"}


class MessageBrokerFactory:
    """QUEUE_SERVICE_TYPE에 따라 MessageBroker 인스턴스를 생성하는 팩토리."""

    @staticmethod
    def create(settings: Settings) -> MessageBroker:
        """환경 변수 기반으로 MessageBroker 인스턴스를 생성한다.

        Args:
            settings: 애플리케이션 설정.

        Returns:
            MessageBroker 구현체 인스턴스.

        Raises:
            ValueError: 지원하지 않는 QUEUE_SERVICE_TYPE인 경우.
        """
        broker_type = settings.QUEUE_SERVICE_TYPE.upper()

        if broker_type == "EVENT_GRID":
            endpoint = getattr(settings, "EVENT_GRID_ENDPOINT", "")
            key = getattr(settings, "EVENT_GRID_KEY", "")
            credential = AzureKeyCredential(key)
            return EventGridAdapter(endpoint=endpoint, credential=credential)

        supported = ", ".join(sorted(SUPPORTED_BROKER_TYPES))
        msg = f"지원하지 않는 QUEUE_SERVICE_TYPE: '{settings.QUEUE_SERVICE_TYPE}'. 지원: {supported}"
        raise ValueError(msg)
